src/tlo/methods/equipment.py

```diff
@@ -0,0 +1,278 @@
+     `master_facilities_list`), an `AssertionError` is raised.
+
+    :param: 'catalogue': The database of all recognised item_codes.
+
+    :param: `data_availability`: Specifies the probability with which each equipment (identified by an `item_code`) is
+     available at a facility level. Note that information is not necessarily provided for every item in the `catalogue`
+     or every facility_id in the `master_facilities_list`.
+
+    :param: `rng`: The Random Number Generator object to use for random numbers.
+
+    :param: `availability`: Determines the mode availability of the equipment. If 'default' then use the availability
+     specified in the `data_availability`; if 'none', then let no equipment be ever be available; if 'all', then all
+     equipment is always available.
+
+    :param `: `master_facilities_list`: The pd.DataFrame with the line-list of all the facilities in the HealthSystem.
+
+    """
+
+    def __init__(
+        self,
+        catalogue: pd.DataFrame,
+        data_availability: pd.DataFrame,
+        rng: np.random,
+        master_facilities_list: pd.DataFrame,
+        availability: Optional[Literal["all", "default", "none"]] = "default",
+    ) -> None:
+        # - Store arguments
+        self.catalogue = catalogue
+        self.rng = rng
+        self.data_availability = data_availability
+        self.availability = availability
+        self.master_facilities_list = master_facilities_list
+
+        # - Data structures for quick look-ups for items and descriptors
+        self._item_code_lookup = self.catalogue.set_index('Item_Description')['Item_Code'].to_dict()
+        self._all_item_descriptors = set(self._item_code_lookup.keys())
+        self._all_item_codes = set(self._item_code_lookup.values())
+        self._all_fac_ids = self.master_facilities_list['Facility_ID'].unique()
+
+        # - Probabilities of items being available at each facility_id
+        self._probabilities_of_items_available = self._calculate_equipment_availability_probabilities()
+
+        # - Internal store of which items have been used at each facility_id This is of the form
+        # {facility_id: {item_code: count}}.
+        self._record_of_equipment_used_by_facility_id = defaultdict(Counter)
+
+
+    def on_simulation_end(self):
+        """Things to do when the simulation ends:
+         * Log (to the summary logger) the equipment that has been used.
+        """
+        self.write_to_log()
+
+    def update_availability(
+        self, availability: Literal["all", "default", "none"]
+    ) -> None:
+        """
+        Update the availability of equipment.
+
+        This is expected to be called midway through the simulation if the
+        assumption of the equipment availability is changed.
+        """
+        assert availability in ["all", "none", "default"], f"New availability parameter {availability} not recognised."
+        self.availability = availability
```

Comment:
The property method seems neater, I agree.

Reply:
implementing that in next commit (hopefully done as intended). In init, I use the setter (`self.availablility = availability`) in order to get the benefit of the checks in the setter (rather than, as you say, storing to the private variable, `self._availability`) -- is that ok?

Reply:
The property implementation looks all correct and you're right that you would set using `self.availablility = availability` to make sure setter method is used rather than using underlying attribute `self._availability`. I was originally thinking here for this to replace the `update_availability` method though maybe that is not necessary as might involve a lot of downstream changes. If keeping the separate `update_availability` method then the separate assertion there is not needed as we will already hit the assert statement in the setter method before changing the value.

Reply:
> In init, I use the setter (self.availablility = availability) in order to get the benefit of the checks in the setter (rather than, as you say, storing to the private variable, self._availability) -- is that ok?

Yep, should behave as intended :+1:

<details>
<summary>(Quick trial by testing)</summary>

```python

class Foo:

    _my_prop: int

    @property
    def my_prop(self) -> int:
        return self._my_prop

    @my_prop.setter
    def my_prop(self, val: int) -> None:
        assert isinstance(val, int) and 0 <= val < 10
        self._my_prop = val

    def __init__(self, prop_val: int = 3):
        self.my_prop = prop_val

f = Foo(prop_val=4) # Sets fine
ff = Foo(prop_val=40) # Assertion error

```
</details>

Reply:
> `update_availability` method though maybe that is not necessary as might involve a lot of downstream changes

`update_availability` used to be [called here](https://github.com/UCL/TLOmodel/blob/3be83633970bd06ee609015c15c901af2b47543e/src/tlo/methods/healthsystem.py#L2781) but now we're referring to the `availability` property directly, we don't seem to need the method? CTRL+F on the codebase doesn't return any hits for `update_availability` anymore.

Reply:
commit just pushed deletes it... 

(Sorry, my mistake. I realised we had deprecated it and used the new setter instead, but forgot to remove it).